### PR TITLE
Update SCAN to handle very large cursor values.

### LINKED ISSUE #2454: Redis scan cursor returns unsigned int, conversion to signed int causes incorrect cursor on following iteration.
### Expected behaviour
When running a redis scan command, I expect the iterator to be the same value as the redis cursor.
### Actual behaviour
The cursor is being converted to a signed int, while redis returns an unsigned int. So any cursor over the signed int max value is set to the signed int max value before the next scan iteration.

### I'm seeing this behaviour on
-OS: Linux
- Redis: 8.1
- PHP: 8.2
- phpredis: 6.0.2

### Steps to reproduce, backtrace or example script
When redis scan command returns a cursor larger than the signed int max value,
$iter will be 9223372036854775807, which is sent to redis for the next iteration.
```
do {
  $arr_keys = $redis->scan($iter, $server, $pattern, $count);
      if($arr_keys !== FALSE){
          foreach($arr_keys as $str_key){
              echo "key";
          }
      }
  echo $iter;
}
while($iter > 0);
```
This is using RedisCluster with a targeted node in the scan parameters

### I've checked
- [x] There is no similar issue from other users
- [x] Issue isn't fixed in `develop` branch

### PR DESCRIPTION
**Pull request by michael-grunder**
Technically Redis may return any unsigned 64 bit integer as a scan cursor.  This presents a problem for PHP in that PHP's integers are signed.  Because of that if a scan cursor is > 2^63 it will overflow and fail to work properly.

This commit updates our SCAN family of commands to deliver cursors in their string form.

```php
public function scan(null|int|string $iterator, ...);
```

On initial entry into our SCAN family we convert either a NULL or empty string cursor to zero, and send the initial scan command.

As Redis replies with cursors we either represent them as a long (if they are <= ZEND_ULONG_MAX) and as a string if greater.  This should mean the fix is minimally breaking as the following code will still work:

```php
$it = NULL;
do {
    print_r($redis->scan($it));
} while ($it !== 0);
```

The `$it !== 0` still works because the zero cursor will be represented as an integer.  Only absurdly large (> 2^63) values are represented as a string.

Fixes #2454

### CODE REVIEW

**Comment by michael-grunder** [comment]
Tested and works on both 32-bit and 64-bit systems.

File: library.c [hunk 1]
```diff
@@ -434,7 +434,7 @@ redis_sock_read_scan_reply(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
     }
 
     /* Push the iterator out to the caller */
-    *iter = atol(p_iter);
+    *cursor = strtoull(p_iter, NULL, 10);
```
Comment (yatsukhnenko) [inline library.c:437]:
```suggestion
    *cursor = ZEND_STRTOUL(p_iter, NULL, 10);
```

File: redis.c [hunk 2]
```diff
@@ -2744,32 +2744,68 @@ redis_build_scan_cmd(char **cmd, REDIS_SCAN_TYPE type, char *key, int key_len,
...
+    }
+}
+
+static uint64_t getScanCursorZVal(zval *zv, zend_bool *was_zero) {
+    ZEND_ASSERT(zv != NULL && (Z_TYPE_P(zv) == IS_LONG ||
+                               Z_TYPE_P(zv) == IS_STRING));;
+
+    if (Z_TYPE_P(zv) == IS_STRING) {
+        *was_zero = Z_STRLEN_P(zv) == 1 && Z_STRVAL_P(zv)[0] == '0';
+        return strtoull(Z_STRVAL_P(zv), NULL, 10);
```
Comment (yatsukhnenko) [inline redis.c:2774]:
```suggestion
        return ZEND_STRTOUL(Z_STRVAL_P(zv), NULL, 10);
```

**Review by yatsukhnenko — APPROVED** [review]

File: library.c (hunk 1 above)
Comment (michael-grunder) [reply library.c:437]: We can't use `ZEND_STRTOUL` here

On a 32-bit system:
```
ZEND_STRTOUL(4890909195324358656) = 4294967295
strtoull(4890909195324358656) = 4890909195324358656
```

Comment (yatsukhnenko) [reply library.c:437]: I guess windows doesn't have `strtoull` that's why I suggested to use zend wrapper but ok, nobody cares about windows :smile:

Comment (michael-grunder) [reply library.c:437]: That's actually a good point.  I checked to make sure that Windows had it but maybe I misunderstood and thought it did.  If we have a build issue I can fix it with a custom `#ifdef`.

**Comment by scottasmith** [comment]
A fix has been merged into Laravel for this into [laravel:11.x](https://github.com/laravel/framework/tree/11.x) with https://github.com/laravel/framework/pull/53095

**Pull request merged**